crates/ruff_python_semantic/src/expressions.rs

```diff
@@ -0,0 +1,65 @@
+use std::ops::{Index, IndexMut};
+
+use ruff_index::{newtype_index, IndexVec};
+use ruff_python_ast::Expr;
+
+/// Id uniquely identifying an expression in a program.
+///
+/// Using a `u32` is sufficient because Ruff only supports parsing documents with a size of max
+/// `u32::max` and it is impossible to have more nodes than characters in the file. We use a
+/// `NonZeroU32` to take advantage of memory layout optimizations.
+#[newtype_index]
+#[derive(Ord, PartialOrd)]
+pub struct ExpressionId;
+
+/// An [`Expr`] AST node in a program, along with a pointer to its parent expression (if any).
+#[derive(Debug)]
+struct Expression<'a> {
```

Comment:
The name here will conflict with `Expr` if we go ahead and rename it to `Expression`. 

Maybe `ExpressionWithParent`. The name doesn't seem as important as it is an internal type.